Display data from internet

minaahilimtiaz opened the pull request.

minaahilimtiaz committed:
Add code to get data from internet

Add api services to get data from internet

Add location manager to get current city

Add location permission checks

Add fragments dynamically

Add data to adapter

Add helper object for formatting functions

AbdulMateenChughtai commented on app/src/main/java/com/example/android/weatherapp/adapter/viewholder/ThreeHourlyWeatherViewHolder.kt:
```diff
@@ -0,0 +1,57 @@
+        setFifthHourlyValues(item[4])
+    }
+
+    private fun setFirstHourlyValues(item: ThreeHourlyWeatherModel) {
```
Create a generic method which takes data and view and updates it accordingly because there's too much repetition of same code here.

minaahilimtiaz replied:
Done. It was visible in next commit

noumanahmad001 commented on app/src/main/java/com/example/android/weatherapp/utilities/Helper.kt:
```diff
@@ -0,0 +1,49 @@
+
+private const val DATE_FORMAT = "MMMM dd, yyyy"
+
+object Helper {
```
Replace object with class

minaahilimtiaz replied:
Created class in the next commit of this branch

noumanahmad001 commented on app/src/main/java/com/example/android/weatherapp/utilities/Helper.kt:
```diff
@@ -0,0 +1,49 @@
+object Helper {
+    const val APP_ID = "4dfdfc7144138f43bfa36b5b3a4b097f"
+
+    fun getDayFromDate(inputDate: String): String{
```
`String{` space issue

minaahilimtiaz replied:
Added space

noumanahmad001 commented on app/src/main/java/com/example/android/weatherapp/utilities/Helper.kt:
```diff
@@ -0,0 +1,49 @@
+    const val APP_ID = "4dfdfc7144138f43bfa36b5b3a4b097f"
+
+    fun getDayFromDate(inputDate: String): String{
+        val format: DateFormat = SimpleDateFormat("yyyy-MM-dd hh:mm:ss", Locale.ENGLISH)
```
`"yyyy-MM-dd hh:mm:ss"` this can be moved to strings ?

minaahilimtiaz replied:
Moved the string to constant file

AbdulMateenChughtai commented on app/src/main/java/com/example/android/weatherapp/utilities/Helper.kt:
```diff
@@ -0,0 +1,49 @@
+        return roundDecimalPlaces(converted)
+    }
+
+    fun roundDecimalPlaces(input: Double): Double{
```
What does this method do?

minaahilimtiaz replied:
It rounds off the value to one decimal place.

noumanahmad001 commented on app/src/main/java/com/example/android/weatherapp/utilities/Helper.kt:
```diff
@@ -0,0 +1,49 @@
+        return date
+    }
+
+    fun convertToCelsius(fahrenheit: Double): Double{
```
spaces issue!

AbdulMateenChughtai commented on app/src/main/java/com/example/android/weatherapp/viewmodel/HomeScreenViewModel.kt:
```diff
@@ -0,0 +1,107 @@
+
+    private fun getForecast() {
+        coroutineScope.launch {
+            val forecastObj = WeatherApi.retrofitService.getFiveDaysWeather("Lahore",
```
Please use repositories.

minaahilimtiaz replied:
Added repositories in next commit of this branch

AbdulMateenChughtai commented on app/src/main/java/com/example/android/weatherapp/viewmodel/HomeScreenViewModel.kt:
```diff
@@ -0,0 +1,107 @@
+
+    private fun prepareWeeklyForecastData(forecastObj: ForecastWeatherModel?) {
+        forecastObj?.let {
+            for (obj in 11 until 40 step 8) {
```
Please avoid using magic numbers.

noumanahmad001 commented on app/src/main/java/com/example/android/weatherapp/viewmodel/HomeScreenViewModel.kt:
```diff
@@ -2,35 +2,96 @@ package com.example.android.weatherapp.viewmodel
-    val currentWeather: LiveData<CurrentWeatherModel>
-        get() = _currentWeather
+    val currentWeather = mutableListOf<CurrentWeatherModel>()
+    val threeHourlyWeatherForecast = mutableListOf<ThreeHourlyWeatherModel>()
```
Weather keyword can be removed from variable name ?

noumanahmad001 commented on app/src/main/java/com/example/android/weatherapp/viewmodel/HomeScreenViewModel.kt:
```diff
@@ -2,35 +2,96 @@ package com.example.android.weatherapp.viewmodel
+    }
+
+    private fun prepareThreeHourlyForecastData(forecastObj: ForecastWeatherModel?) {
+        forecastObj?.let {
```
why Obj ?

noumanahmad001 commented on app/src/main/java/com/example/android/weatherapp/viewmodel/HomeScreenViewModel.kt:
```diff
@@ -2,35 +2,96 @@ package com.example.android.weatherapp.viewmodel
+
+    private fun getForecast() {
+        coroutineScope.launch {
+            val forecastObj = WeatherApi.retrofitService.getFiveDaysWeather("Lahore",
```
No Obj !!

noumanahmad001 commented on app/src/main/java/com/example/android/weatherapp/viewmodel/HomeScreenViewModel.kt:
```diff
@@ -2,35 +2,96 @@ package com.example.android.weatherapp.viewmodel
         coroutineScope.launch {
-            var weatherObj = WeatherApi.retrofitService.getCurrentWeather("London", APP_ID)
+            val weatherObj = WeatherApi.retrofitService.getCurrentWeather("Lahore",
+                Helper.APP_ID)
```
constants should be in separate file or strings.xml

minaahilimtiaz replied:
Moved to separate constant file

AbdulMateenChughtai commented on app/src/main/res/layout/three_hourly_weather_layout.xml:
```diff
@@ -0,0 +1,58 @@
+        layout="@layout/hourly_weather_layout"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="160dp"
```
Use chaining instead of relying on margin start only.

noumanahmad001 commented on app/src/main/java/com/example/android/weatherapp/viewmodel/HomeScreenViewModel.kt:
```diff
@@ -2,35 +2,96 @@ package com.example.android.weatherapp.viewmodel
+            val forecastObj = WeatherApi.retrofitService.getFiveDaysWeather("Lahore",
+                Helper.APP_ID)
+            try {
+                if(forecastObj.isSuccessful) {
```
Recheck logic of this code.

noumanahmad001 commented on app/src/main/java/com/example/android/weatherapp/viewmodel/HomeScreenViewModel.kt:
```diff
@@ -2,35 +2,96 @@ package com.example.android.weatherapp.viewmodel
+
+    private fun prepareWeeklyForecastData(forecastObj: ForecastWeatherModel?) {
+        forecastObj?.let {
+            for (obj in 11 until 40 step 8) {
```
No obj  !!!

minaahilimtiaz committed:
Add repository in the architecture

Add repository for accessing data from source

Add progress bar

Add text view for displaying error

Improve code quality

AbdulMateenChughtai commented on app/src/main/java/com/example/android/weatherapp/WeatherRepository.kt:
```diff
@@ -0,0 +1,91 @@
+
+    private fun prepareThreeHourlyForecastData(result: ForecastWeatherModel?) {
+        result?.let {
+            val tempList = result.list.subList(START_INDEX,END_INDEX)
```
Return if result is null and try maybe removing this temporary variable.

AbdulMateenChughtai commented on app/src/main/java/com/example/android/weatherapp/WeatherRepository.kt:
```diff
@@ -0,0 +1,91 @@
+class WeatherRepository {
+
+    private val weatherForecast: ForecastData = ForecastData()
+    private val helper: Helper = Helper()
```
Rename Helper to some more descriptive name like WeatherHelper maybe?

minaahilimtiaz replied:
Renamed it to **formatHelper**

AbdulMateenChughtai commented on app/src/main/java/com/example/android/weatherapp/WeatherRepository.kt:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    private fun prepareWeeklyForecastData(result: ForecastWeatherModel?) {
+        result?.let {
```
Return if result is null.

minaahilimtiaz replied:
Done. Have added the null check

AbdulMateenChughtai commented on app/src/main/java/com/example/android/weatherapp/WeatherRepository.kt:
```diff
@@ -0,0 +1,91 @@
+            } catch (e: HttpException) {
+                println("Exception1  ${e.message}")
+            } catch (e: Throwable) {
+                println("Repo1 + ${e.message}")
```
Try handling errors more gracefully like showing a Banner in case of error.

AbdulMateenChughtai commented on app/src/main/java/com/example/android/weatherapp/activity/BaseActivity.kt:
```diff
@@ -2,6 +2,6 @@ package com.example.android.weatherapp.activity
 import androidx.appcompat.app.AppCompatActivity
 
-open class BaseActivity() : AppCompatActivity(){
+open class BaseActivity() : AppCompatActivity() {
```
v nice.

AbdulMateenChughtai commented on app/src/main/java/com/example/android/weatherapp/fragment/HomeScreenFragment.kt:
```diff
@@ -26,16 +28,65 @@ class HomeScreenFragment : BaseFragment() {
     }
+
+    private fun setOnLoadingObserver(binding: FragmentHomeScreenBinding) {
+        viewModel.eventLoading.observe(viewLifecycleOwner, Observer {hasLoaded ->
```
Add a space before hasLoaded.

minaahilimtiaz replied:
Done. It will be visible in next commit

AbdulMateenChughtai commented on app/src/main/java/com/example/android/weatherapp/fragment/HomeScreenFragment.kt:
```diff
@@ -26,16 +28,65 @@ class HomeScreenFragment : BaseFragment() {
+
+    private fun setOnLoadingObserver(binding: FragmentHomeScreenBinding) {
+        viewModel.eventLoading.observe(viewLifecycleOwner, Observer {hasLoaded ->
+            if(hasLoaded) {
```
Make a separate method in base fragment to handle this if-else logic, so that your observer method remains clean.

AbdulMateenChughtai commented on app/src/main/java/com/example/android/weatherapp/models/ThreeHourlyWeatherModel.kt:
```diff
@@ -1,6 +1,6 @@
 package com.example.android.weatherapp.models
 
-data class ThreeHourlyWeatherModel (
+data class ThreeHourlyWeatherModel(
```
Add a space before parenthesis.

minaahilimtiaz replied:
https://kotlinlang.org/docs/reference/data-classes.html
here there is no space before parenthesis. Should i still add the space?

AbdulMateenChughtai commented on app/src/main/java/com/example/android/weatherapp/viewmodel/HomeScreenViewModel.kt:
```diff
@@ -2,106 +2,60 @@ package com.example.android.weatherapp.viewmodel
-               println("Exception ${e.message}")
-            } catch (e: Throwable) {
-                println("Ooops: Something else went wrong")
+            try{
```
Add a space after try.

minaahilimtiaz replied:
Done. It will be visible in next commit

minaahilimtiaz committed:
Improve code quality

minaahilimtiaz committed:
Add simple error handling

AbdulMateenChughtai commented on app/src/main/java/com/example/android/weatherapp/WeatherRepository.kt:
```diff
@@ -6,10 +6,10 @@ import kotlinx.coroutines.Dispatchers
 
     private val weatherForecast: ForecastData = ForecastData()
-    private val helper: Helper = Helper()
+    private val formatHelper: Helper = Helper()
```
I meant rename the Helper class itself, not its instance.

AbdulMateenChughtai commented on app/src/main/java/com/example/android/weatherapp/fragment/BaseFragment.kt:
```diff
@@ -1,7 +1,57 @@
+        }
+    }
+
+    fun checkLoading(isLoaded: Boolean, progressBar: ProgressBar) {
```
checkLoading is not an appropriate name. Naming things is important as it makes the code more readable.

minaahilimtiaz committed:
Merge branch 'Developer' into display-data-from-internet
